gencap.f90

```diff
@@ -12,324 +12,468 @@
-    return
-    end if
+      if (nq*nv .ne. 0) then
+      print*, "Oh no! nq and nv can't both be nonzero."
+      return
+      end if
 
-    if (.not. allocated(tab_r)) then !
-        print*,"Errorface of errors: you haven't called captn_init to load the solar model!"
-        return
-    end if
-        allocate(u_int_res(nlines))
+      if (.not. allocated(tab_r)) then
+        stop "You haven't yet called captn_init to load the solar model!"
+      end if
+      allocate(u_int_res(nlines))
 
-!    As far as I can tell, the second argument (fofuoveru) does nothing in this integrator. I've sent it to an inert dummy just in case.
-    capped = 0.d0
+      capped = 0.d0
 
-    do ri=1,nlines !loop over the star
-    result = 0.d0
-    ri_for_omega = ri !accessed via the module
-    !call integrator
-    call dsntdqagse(integrand,dummyf,1.d0,vesc_halo, &
-    epsabs,epsrel,limit,result,abserr,neval,ier,alist,blist,rlist,elist,iord,last)
+      !Loop over the shells of constant radius in the star
+      do ri = 1, nlines
 
+        vesc = tab_vesc(ri)
+        vesc_shared = vesc !make accessible via the module
 
+        !Loop over the different elements
+        do eli = 1, niso
 
-    u_int_res(ri) = result*sigma_0
-    capped = capped + tab_r(ri)**2*u_int_res(ri)*tab_dr(ri)
+          a = AtomicNumber(eli)
+          a_shared = a !make accessible via the module
 
-    end do
-    capped = 4.d0*pi*Rsun**3*capped
+          !This is fine for SD as long as it's just hydrogen. Otherwise, spins must be added.
+          sigma_N = a**2 * (sigma_SI*a**2 + sigma_SD) * (mx_in+mnuc)**2/(mx_in+a*mnuc)**2
 
-    if (capped .gt. 1.d100) then
-      print*,"Capt'n General says: Oh my, it looks like you are capturing an  &
-      infinite amount of dark matter in the Sun. Best to look into that."
-    end if
+          mu = mx_in/(mnuc*a)
+          muplus = (1.+mu)/2.
+          muminus = (mu-1.d0)/2.
 
-    !this now has its own function:
-    ! maxcap = pi/3.d0*rho0/mdm*Rsun**2 &
-    ! *(exp(-3./2.*usun**2/u0**2)*sqrt(6.d0/pi)*u0 &
-    ! + (6.d0*GMoverR/usun + (u0**2 + 3.d0*usun**2)/usun)*erf(sqrt(3./2.)*usun/u0))
+          ! Bottom part of the integral is always zero -- happy little slow DM particles can always be captured.
+          umin = 0.d0
+          ! Chop the top of the integral off at the smaller of the halo escape velocity or the minimum velocity required for capture.
+          umax = min(vesc * sqrt(mu)/abs(muminus), vesc_halo)
 
-  !  print*,"sigma_0 =", sigma_0, "; m = ", mdm, "; nq = ", nq, "; Capture rate: ", capped, "max = ", maxcap
+          !Call integrator
+          call dsntdqagse(integrand,vdist_over_u,umin,umax, &
+          epsabs,epsrel,limit,u_int_res(ri),abserr,neval,ier,alist,blist,rlist,elist,iord,last)
+          u_int_res(ri) = u_int_res(ri) * 2.d0 * sigma_N * NAvo * tab_starrho(ri)*tab_mfr(ri,eli) * (muplus/mx_in)**2
+          capped = capped + tab_r(ri)**2*u_int_res(ri)*tab_dr(ri)
```

Comment:
Why is the result of the integrator stored in one index of an array, can u_int_res just be a double type?